clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -164,22 +167,29 @@ public synchronized AuthenticationException getAndClearAuthenticationException()
      * Wait for metadata update until the current version is larger than the last version we know of
      */
     public synchronized void awaitUpdate(final int lastVersion, final long maxWaitMs) throws InterruptedException {
-        if (maxWaitMs < 0) {
+        if (maxWaitMs < 0)
             throw new IllegalArgumentException("Max time to wait for metadata updates should not be < 0 milliseconds");
-        }
+
         long begin = System.currentTimeMillis();
         long remainingWaitMs = maxWaitMs;
-        while (this.version <= lastVersion) {
-            AuthenticationException ex = getAndClearAuthenticationException();
-            if (ex != null)
-                throw ex;
-            if (remainingWaitMs != 0)
-                wait(remainingWaitMs);
-            long elapsed = System.currentTimeMillis() - begin;
-            if (elapsed >= maxWaitMs)
-                throw new TimeoutException("Failed to update metadata after " + maxWaitMs + " ms.");
-            remainingWaitMs = maxWaitMs - elapsed;
+        try {
+            while ((this.version <= lastVersion) && !isClosed()) {
+                AuthenticationException ex = getAndClearAuthenticationException();
+                if (ex != null)
+                    throw ex;
+                if (remainingWaitMs != 0)
+                    wait(remainingWaitMs);
+                long elapsed = System.currentTimeMillis() - begin;
+                if (elapsed >= maxWaitMs)
+                    throw new TimeoutException("Failed to update metadata after " + maxWaitMs + " ms.");
+                remainingWaitMs = maxWaitMs - elapsed;
+            }
+        } catch (InterruptedException e) {
```

Comment:
I am rethrowing as `KafkaException` if the interrupt was because of producer close; `close()` calls `notifyAll()` which could interrupt the `wait()` in this method. Does this seem reasonable?

Reply:
The problem is that we are losing the indication that the interrupt has occurred. A caller up the stack may depend on seeing it. I think I would just let the exception be raised in all cases even if the producer is being closed.